### PR TITLE
ci: add permissions block to workflows

### PR DESCRIPTION
**Pull request by reakaleek**
## Details

⚠️ This PR was created by an automated tool. Please review the changes carefully. ⚠️ 

We want to set the default permissions for workflows to read-only for contents. 
This is a security measure to prevent accidental changes to the repository.

This change adds a top-level permissions block to all workflows in the .github/workflows directory.
```yaml
permissions:
  contents: read
```

In some cases workflows might need more permissions than just `contents: read`.
Please checkout this branch and add the necessary permissions to the workflows.

If your workflow uses a Personal Access Token (PAT), we can still add the permissions block,
   but it will not have any effect.

Merging this PR as is might cause workflows that need more permissions to fail.

If there are any questions, please reach out to the @elastic/observablt-ci

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -6,6 +6,9 @@ on:
     branches:
       - main
 
+permissions:
+  contents: read
```
Comment (reakaleek) [inline .github/workflows/ci.yml:10]:
```suggestion
  contents: read
  actions: write
```

File: .github/workflows/e2e.yml [hunk 2]
```diff
@@ -8,6 +8,9 @@ on:
     branches:
       - main
 
+permissions:
+  contents: read
```
Comment (reakaleek) [inline .github/workflows/e2e.yml:12]:
```suggestion
  contents: read
  actions: write
```

For uploading artifacts

**Review by v1v — DISMISSED** [review]

File: .github/workflows/test-reporter.yml [hunk 3]
```diff
@@ -9,6 +9,10 @@ on:
     types:
       - completed
 
+permissions:
+  contents: read
+  actions: read
```
Comment (reakaleek) [inline .github/workflows/test-reporter.yml:14]:
```suggestion
  actions: read
  checks: write
```

File: .github/workflows/e2e.yml [hunk 4]
```diff
@@ -8,6 +8,11 @@ on:
...
+  actions: write
```
Comment (reakaleek) [inline .github/workflows/e2e.yml:13]:
```suggestion
```

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -6,6 +6,11 @@ on:
...
+  actions: write
```
Comment (reakaleek) [inline .github/workflows/ci.yml:11]:
```suggestion
```

**Review by v1v — DISMISSED** [review]

File: .github/workflows/e2e.yml [hunk 6]
```diff
@@ -8,6 +8,10 @@ on:
...
```
Comment (reakaleek) [inline .github/workflows/e2e.yml:13]:
```suggestion
```

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -6,6 +6,10 @@ on:
...
```
Comment (reakaleek) [inline .github/workflows/ci.yml:11]:
```suggestion
```

**Review by v1v — APPROVED** [review]

**Pull request merged**